File: ycmd/completers/all/tests/identifier_completer_test.py
#!/usr/bin/env python
#
# Copyright (C) 2013  Google Inc.
#
# This file is part of YouCompleteMe.
#
# YouCompleteMe is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# YouCompleteMe is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.

from nose.tools import eq_
from ycmd.completers.all import identifier_completer as ic
from ycmd.request_wrap import RequestWrap
from ycmd.tests.test_utils import BuildRequest

def BuildRequestWrap( contents, column_num, line_num = 1 ):
  return RequestWrap( BuildRequest( column_num = column_num,
                                    line_num = line_num,
                                    contents = contents ) )


def GetCursorIdentifier_StartOfLine_test():
  eq_( 'foo', ic._GetCursorIdentifier( BuildRequestWrap( 'foo', 1 ) ) )
  eq_( 'fooBar', ic._GetCursorIdentifier( BuildRequestWrap( 'fooBar', 1 ) ) )


def GetCursorIdentifier_EndOfLine_test():
  eq_( 'foo', ic._GetCursorIdentifier( BuildRequestWrap( 'foo', 3 ) ) )


def GetCursorIdentifier_PastEndOfLine_test():
  eq_( '', ic._GetCursorIdentifier( BuildRequestWrap( 'foo', 11 ) ) )


def GetCursorIdentifier_NegativeColumn_test():
  eq_( 'foo', ic._GetCursorIdentifier( BuildRequestWrap( 'foo', -10 ) ) )


def GetCursorIdentifier_StartOfLine_StopsAtNonIdentifierChar_test():
  eq_( 'foo', ic._GetCursorIdentifier( BuildRequestWrap( 'foo(goo)', 1 ) ) )


def GetCursorIdentifier_AtNonIdentifier_test():
  eq_( 'goo', ic._GetCursorIdentifier( BuildRequestWrap( 'foo(goo)', 4 ) ) )


def GetCursorIdentifier_WalksForwardForIdentifier_test():
  eq_( 'foo', ic._GetCursorIdentifier( BuildRequestWrap( '       foo', 1 ) ) )


def GetCursorIdentifier_FindsNothingForward_test():
  eq_( '', ic._GetCursorIdentifier( BuildRequestWrap( 'foo   ()***()', 5 ) ) )


def GetCursorIdentifier_SingleCharIdentifier_test():
  eq_( 'f', ic._GetCursorIdentifier( BuildRequestWrap( '    f    ', 1 ) ) )


def GetCursorIdentifier_StartsInMiddleOfIdentifier_test():
  eq_( 'foobar', ic._GetCursorIdentifier( BuildRequestWrap( 'foobar', 4 ) ) )


def GetCursorIdentifier_LineEmpty_test():
  eq_( '', ic._GetCursorIdentifier( BuildRequestWrap( '', 12 ) ) )


def PreviousIdentifier_Simple_test():
  eq_( 'foo', ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo', 4 ) ) )


def PreviousIdentifier_ColumnInMiddleStillWholeIdent_test():
  eq_( 'foobar', ic._PreviousIdentifier( 2, BuildRequestWrap( 'foobar', 4 ) ) )


def PreviousIdentifier_IgnoreForwardIdents_test():
  eq_( 'foo',
       ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo bar zoo', 4 ) ) )


def PreviousIdentifier_IgnoreTooSmallIdent_test():
  eq_( '', ic._PreviousIdentifier( 4, BuildRequestWrap( 'foo', 4 ) ) )


def PreviousIdentifier_IgnoreTooSmallIdent_DontContinueLooking_test():
  eq_( '', ic._PreviousIdentifier( 4, BuildRequestWrap( 'abcde foo', 10 ) ) )


def PreviousIdentifier_WhitespaceAfterIdent_test():
  eq_( 'foo', ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo     ', 6 ) ) )


def PreviousIdentifier_JunkAfterIdent_test():
  eq_( 'foo',
       ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo  ;;()**   ', 13 ) ) )


def PreviousIdentifier_IdentInMiddleOfJunk_test():
  eq_( 'aa',
       ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo  ;;(aa)**   ', 13 ) ) )


def PreviousIdentifier_IdentOnPreviousLine_test():
  eq_( 'foo',
       ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo\n   ',
                                                    column_num = 3,
                                                    line_num = 2 ) ) )

  eq_( 'foo',
       ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo\n',
                                                    column_num = 1,
                                                    line_num = 2 ) ) )


def PreviousIdentifier_IdentOnPreviousLine_JunkAfterIdent_test():
  eq_( 'foo',
       ic._PreviousIdentifier( 2, BuildRequestWrap( 'foo **;()\n   ',
                                                    column_num = 3,
                                                    line_num = 2 ) ) )

File: cpp/ycm/tests/IdentifierCompleter_test.cpp
// Copyright (C) 2011-2018 ycmd contributors
//
// This file is part of ycmd.
//
// ycmd is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// ycmd is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with ycmd.  If not, see <http://www.gnu.org/licenses/>.

#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include "IdentifierCompleter.h"
#include "Utils.h"
#include "TestUtils.h"

using ::testing::ElementsAre;
using ::testing::IsEmpty;
using ::testing::WhenSorted;

namespace YouCompleteMe {

TEST( IdentifierCompleterTest, SortOnEmptyQuery ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foo",
                 "bar" } ).CandidatesForQuery( "" ),
               ElementsAre( "bar",
                            "foo" ) );
}

TEST( IdentifierCompleterTest, IgnoreEmptyCandidate ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "" } ).CandidatesForQuery( "" ),
               IsEmpty() );
}

TEST( IdentifierCompleterTest, IgnoreCandidatesShorterThanQuery ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "fo",
                 "foo" } ).CandidatesForQuery( "fooo" ),
               IsEmpty() );
}

TEST( IdentifierCompleterTest, NoDuplicatesReturned ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "foobar",
                 "foobar" } ).CandidatesForQuery( "foo" ),
               ElementsAre( "foobar" ) );
}


TEST( IdentifierCompleterTest, OneCandidate ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar" } ).CandidatesForQuery( "fbr" ),
               ElementsAre( "foobar" ) );
}

TEST( IdentifierCompleterTest, ManyCandidateSimple ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "foobartest",
                 "Foobartest" } ).CandidatesForQuery( "fbr" ),
               WhenSorted( ElementsAre( "Foobartest",
                                        "foobar",
                                        "foobartest" ) ) );
}

TEST( IdentifierCompleterTest, SmartCaseFiltering ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "fooBar",
                 "fooBaR" } ).CandidatesForQuery( "fBr" ),
               ElementsAre( "fooBaR",
                            "fooBar" ) );
}

TEST( IdentifierCompleterTest, FirstCharSameAsQueryWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "afoobar" } ).CandidatesForQuery( "fbr" ),
               ElementsAre( "foobar",
                            "afoobar" ) );
}

TEST( IdentifierCompleterTest, CompleteMatchForWordBoundaryCharsWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "FooBarQux",
                 "FBaqux" } ).CandidatesForQuery( "fbq" ),
               ElementsAre( "FooBarQux",
                            "FBaqux" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "CompleterTest",
                 "CompleteMatchForWordBoundaryCharsWins" } )
               .CandidatesForQuery( "ct" ),
               ElementsAre( "CompleterTest",
                            "CompleteMatchForWordBoundaryCharsWins" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "FooBar",
                 "FooBarRux" } ).CandidatesForQuery( "fbr" ),
               ElementsAre( "FooBarRux",
                            "FooBar" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "foo-bar",
                 "foo-bar-rux" } ).CandidatesForQuery( "fbr" ),
               ElementsAre( "foo-bar-rux",
                            "foo-bar" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "foo.bar",
                 "foo.bar.rux" } ).CandidatesForQuery( "fbr" ),
               ElementsAre( "foo.bar.rux",
                            "foo.bar" ) );
}

TEST( IdentifierCompleterTest, RatioUtilizationTieBreak ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "aGaaFooBarQux",
                 "aBaafbq" } ).CandidatesForQuery( "fbq" ),
               ElementsAre( "aGaaFooBarQux",
                            "aBaafbq" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "aFooBarQux",
                 "afbq" } ).CandidatesForQuery( "fbq" ),
               ElementsAre( "aFooBarQux",
                            "afbq" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "acaaCaaFooGxx",
                 "aCaafoog" } ).CandidatesForQuery( "caafoo" ),
               ElementsAre( "acaaCaaFooGxx",
                            "aCaafoog" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "FooBarQux",
                 "FooBarQuxZaa" } ).CandidatesForQuery( "fbq" ),
               ElementsAre( "FooBarQux",
                            "FooBarQuxZaa" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "FooBar",
                 "FooBarRux" } ).CandidatesForQuery( "fba" ),
               ElementsAre( "FooBar",
                            "FooBarRux" ) );
}

TEST( IdentifierCompleterTest, QueryPrefixOfCandidateWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "fbaroo" } ).CandidatesForQuery( "foo" ),
               ElementsAre( "foobar",
                            "fbaroo" ) );
}

TEST( IdentifierCompleterTest, LowerMatchCharIndexSumWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "ratio_of_word_boundary_chars_in_query_",
                 "first_char_same_in_query_and_text_"
               } ).CandidatesForQuery( "charinq" ),
               ElementsAre( "first_char_same_in_query_and_text_",
                            "ratio_of_word_boundary_chars_in_query_" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "barfooq",
                 "barquxfoo" } ).CandidatesForQuery( "foo" ),
               ElementsAre( "barfooq",
                            "barquxfoo" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "xxxxxxabc",
                 "xxabcxxxx" } ).CandidatesForQuery( "abc" ),
               ElementsAre( "xxabcxxxx",
                            "xxxxxxabc" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "FooBarQux",
                 "FaBarQux" } ).CandidatesForQuery( "fbq" ),
               ElementsAre( "FaBarQux",
                            "FooBarQux" ) );
}

TEST( IdentifierCompleterTest, ShorterCandidateWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "CompleterT",
                 "CompleterTest" } ).CandidatesForQuery( "co" ),
               ElementsAre( "CompleterT",
                            "CompleterTest" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "CompleterT",
                 "CompleterTest" } ).CandidatesForQuery( "plet" ),
               ElementsAre( "CompleterT",
                            "CompleterTest" ) );
}

TEST( IdentifierCompleterTest, SameLowercaseCandidateWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "Foobar" } ).CandidatesForQuery( "foo" ),
               ElementsAre( "foobar",
                            "Foobar" ) );

}

TEST( IdentifierCompleterTest, PreferLowercaseCandidate ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "chatContentExtension",
                 "ChatContentExtension"
               } ).CandidatesForQuery( "chatContent" ),
               ElementsAre( "chatContentExtension",
                            "ChatContentExtension" ) );

  EXPECT_THAT( IdentifierCompleter( {
                 "fooBar",
                 "FooBar" } ).CandidatesForQuery( "oba" ),
               ElementsAre( "fooBar",
                            "FooBar" ) );
}

TEST( IdentifierCompleterTest, ShorterAndLowercaseWins ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "STDIN_FILENO",
                 "stdin" } ).CandidatesForQuery( "std" ),
               ElementsAre( "stdin",
                            "STDIN_FILENO" ) );
}


TEST( IdentifierCompleterTest, NonAlnumChars ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "font-family",
                 "font-face" } ).CandidatesForQuery( "fo" ),
               ElementsAre( "font-face",
                            "font-family" ) );
}


TEST( IdentifierCompleterTest, NonAlnumStartChar ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "-zoo-foo" } ).CandidatesForQuery( "-z" ),
               ElementsAre( "-zoo-foo" ) );
}


TEST( IdentifierCompleterTest, UnicodeCandidates ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "uni¢𐍈d€" } ).CandidatesForQuery( "¢" ),
               ElementsAre( "uni¢𐍈d€" ) );
}


TEST( IdentifierCompleterTest, NonPrintableCandidates ) {
  EXPECT_THAT( IdentifierCompleter( {
                 "\x01\x1f\x7f" } ).CandidatesForQuery( "\x1f" ),
               ElementsAre( "\x01\x1f\x7f" ) );
}


TEST( IdentifierCompleterTest, LotOfCandidates ) {
  // Generate a lot of candidates of the form [a-z]{5} in reverse order.
  std::vector< std::string > candidates;
  for ( int i = 0; i < 2048; ++i ) {
    std::string candidate = "";
    int letter = i;
    for ( int pos = 0; pos < 5; letter /= 26, ++pos ) {
      candidate = std::string( 1, letter % 26 + 'a' ) + candidate;
    }
    candidates.insert( candidates.begin(), candidate );
  }

  IdentifierCompleter completer( candidates );

  std::reverse( candidates.begin(), candidates.end() );

  EXPECT_THAT( completer.CandidatesForQuery( "aa" ),
               candidates );

  EXPECT_THAT( completer.CandidatesForQuery( "aa", 2 ),
               ElementsAre( "aaaaa",
                            "aaaab" ) );
}


TEST( IdentifierCompleterTest, TagsEndToEndWorks ) {
  IdentifierCompleter completer;
  std::vector< std::string > tag_files;
  tag_files.push_back( PathToTestFile( "basic.tags" ).string() );

  completer.AddIdentifiersToDatabaseFromTagFiles( tag_files );

  EXPECT_THAT( completer.CandidatesForQueryAndType( "fo", "cpp" ),
               ElementsAre( "foosy",
                            "fooaaa" ) );

}


// Filetype checking
TEST( IdentifierCompleterTest, ManyCandidateSimpleFileType ) {
  IdentifierCompleter completer;
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "foobartest",
                 "Foobartest"
               }, "c", "foo" ).CandidatesForQueryAndType( "fbr", "c" ),
               WhenSorted( ElementsAre( "Foobartest",
                                        "foobar",
                                        "foobartest" ) ) );
}


TEST( IdentifierCompleterTest, ManyCandidateSimpleWrongFileType ) {
  IdentifierCompleter completer;
  EXPECT_THAT( IdentifierCompleter( {
                 "foobar",
                 "foobartest",
                 "Foobartest"
               }, "c", "foo" ).CandidatesForQueryAndType( "fbr", "cpp" ),
               IsEmpty() );
}

} // namespace YouCompleteMe

